tuna/connection.py

```diff
@@ -190,87 +276,46 @@ def exec_command(self, cmd, timeout=SSH_TIMEOUT, abort=None, proc_line=None):
       if not proc_line:
         # pylint: disable-next=unnecessary-lambda-assignment ; more readable
         proc_line = lambda x: self.logger.info(line.strip())
-      ret_out = StringIO()
-      ret_err = e_var
+      ret_out: Union[TextIO, StringIO] = StringIO()
+      ret_err: ChannelStderrFile = e_var
+      ret_code: int = 0
       try:
         while True:
-          line = o_var.readline()
+          line: str = o_var.readline()
           if line == '' and not self.is_alive():
             break
           if line:
             proc_line(line)
             ret_out.write(line)
         ret_out.seek(0)
         if self.local_machine:
-          ret_code = self.subp.returncode
+          if self.subp is not None:
+            ret_code = self.subp.returncode
+          else:
+            ret_code = 0
           if ret_out:
             ret_out.seek(0)
-            ret_err = StringIO()
-            err_str = ret_out.readlines()
+            ret_err = StringIO()  #type: ignore
+            err_str: list = ret_out.readlines()
             for line in err_str[-5:]:
               ret_err.write(line)
             ret_err.seek(0)
             ret_out.seek(0)
         else:
-          ret_code = self.out_channel.recv_exit_status()
+          if self.out_channel is not None and ret_code is not None:
+            ret_code = self.out_channel.recv_exit_status()
+
       except (socket.timeout, socket.error) as exc:
         self.logger.warning('Exception occurred %s', exc)
         ret_code = 1
-        ret_out = None
```

Comment:
This changes behavior. Might be ok, but I want to know how it affects the methods that use this function. May need to be an update to read the ret_code first.